opt: cache restricted interesting orderings

mgartner opened the pull request:
#### opt: cache simplified interesting orderings

The `GenerateMergeJoins` rule simplifies the interesting orderings of a
join's left child with respect to the child's functional dependencies.
Simplifying an ordering set is computationally expensive, and it can
dominate the CPU-time of optimizing queries with many joins and
interesting orderings. In a query with many joins, `GenerateMergeJoins`
fires repeatedly with the same left child. To avoid repeatedly computing
the same simplified interesting orderings, we cache them in
`props.Relational.Rule.SimplifiedInterestingOrderings`. This
significantly reduces the planning time of some queries.

Release note (performance improvement): Query planning time has been reduced
for some queries with multiple joins.

#### opt: cache restricted interesting orderings

This commit builds on the last commit by caching interesting orderings
that are restricted to a specific column set, not just simplified
interesting orderings. Restricting an ordering set is computationally
expensive (see `OrderingSet.RestrictToCols`), and queries with many
joins can repeatedly restrict the same interesting orderings of a
relation with the same set of columns. Caching restricted orderings
eliminates this duplicate work.

Epic: None

Release note: None

mgartner committed:
opt: cache simplified interesting orderings

The `GenerateMergeJoins` rule simplifies the interesting orderings of a
join's left child with respect to the child's functional dependencies.
Simplifying an ordering set is computationally expensive, and it can
dominate the CPU-time of optimizing queries with many joins and
interesting orderings. In a query with many joins, `GenerateMergeJoins`
fires repeatedly with the same left child. To avoid repeatedly computing
the same simplified interesting orderings, we cache them in
`props.Relational.Rule.SimplifiedInterestingOrderings`. This
significantly reduces the planning time of some queries.

Release note (performance improvement): Query planning time has been reduced
for some queries with multiple joins.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/102011)
<!-- Reviewable:end -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great improvement!!

> Reviewed 14 of 14 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102011#-:-NT_2YSQE9fi3YRdAVcc:b-m3m9jg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @rharding6373)

___
*[`pkg/sql/opt/ordering/interesting_orderings.go` line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102011#-NT_0mFx8VOdXtfktfpg:-NT_0mFx8VOdXtfktfph:b-ot87hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/779ba3f488d21dbd8081de279d3b7ef404cb4505/pkg/sql/opt/ordering/interesting_orderings.go#L32)):*
> ```Go
> 
> 	// Derive the interesting orderings and simplify them.
> 	orders := DeriveInterestingOrderings(e).Copy()
> ```

Could we do even better by first checking whether any of the orders needs to be simplified? Or does that just not have an impact?
___
*[`pkg/sql/opt/props/logical.go` line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102011#-NT_22iDCcAcu4PeixXl:-NT_22iDCcAcu4PeixXm:b-qwn00s) ([raw file](https://github.com/cockroachdb/cockroach/blob/779ba3f488d21dbd8081de279d3b7ef404cb4505/pkg/sql/opt/props/logical.go#L274)):*
> ```Go
> 		// relational expression's functional dependencies. See
> 		// OrderingChoice.Simplify.
> 		SimplifiedInterestingOrderings OrderingSet
> ```

What's preventing us from simplifying the `InterestingOrderings` field instead of adding a new one?


<!-- Sent from Reviewable.io -->

mgartner committed:
opt: cache restricted interesting orderings

This commit builds on the last commit by caching interesting orderings
that are restricted to a specific column set, not just simplified
interesting orderings. Restricting an ordering set is computationally
expensive (see `OrderingSet.RestrictToCols`), and queries with many
joins can repeatedly restrict the same interesting orderings of a
relation with the same set of columns. Caching restricted orderings
eliminates this duplicate work.

Release note: None

mgartner commented:
I've added a new commit that makes further improvements. The benchmarks below compare the results of `master` to the results when both of these commits are applied:

```
name                         old time/op    new time/op    delta
SlowQueries/slow-query-1-10    15.9ms ± 1%    15.3ms ± 1%   -3.63%  (p=0.008 n=5+5)
SlowQueries/slow-query-2-10     221ms ± 0%     218ms ± 0%   -1.38%  (p=0.016 n=5+4)
SlowQueries/slow-query-3-10    62.9ms ± 0%    62.7ms ± 0%     ~     (p=0.310 n=5+5)
SlowQueries/slow-query-4-10     1.38s ± 0%     0.36s ± 1%  -73.70%  (p=0.008 n=5+5)

name                         old alloc/op   new alloc/op   delta
SlowQueries/slow-query-1-10    7.21MB ± 0%    7.02MB ± 0%   -2.68%  (p=0.008 n=5+5)
SlowQueries/slow-query-2-10    49.4MB ± 0%    49.4MB ± 0%     ~     (p=0.548 n=5+5)
SlowQueries/slow-query-3-10    44.8MB ± 0%    44.8MB ± 0%   +0.07%  (p=0.008 n=5+5)
SlowQueries/slow-query-4-10     737MB ± 0%     195MB ± 0%  -73.50%  (p=0.008 n=5+5)

name                         old allocs/op  new allocs/op  delta
SlowQueries/slow-query-1-10     76.7k ± 0%     70.1k ± 0%   -8.57%  (p=0.008 n=5+5)
SlowQueries/slow-query-2-10      403k ± 0%      393k ± 0%   -2.47%  (p=0.008 n=5+5)
SlowQueries/slow-query-3-10      389k ± 0%      389k ± 0%   -0.09%  (p=0.008 n=5+5)
SlowQueries/slow-query-4-10     17.4M ± 0%      3.2M ± 0%  -81.76%  (p=0.008 n=5+5)
```

```
name                                                    old time/op    new time/op    delta
Phases/kv-read/Simple/Explore-10                          26.2µs ± 1%    26.1µs ± 1%    ~     (p=0.222 n=5+5)
Phases/kv-read-const/Simple/Explore-10                    26.2µs ± 0%    26.2µs ± 0%    ~     (p=0.730 n=5+4)
Phases/tpcc-new-order/Simple/Explore-10                   72.7µs ± 0%    72.6µs ± 1%    ~     (p=0.548 n=5+5)
Phases/tpcc-delivery/Simple/Explore-10                    55.8µs ± 0%    55.6µs ± 0%    ~     (p=0.222 n=5+5)
Phases/tpcc-delivery/Prepared/Explore-10                  18.8µs ± 3%    18.4µs ± 1%  -2.04%  (p=0.008 n=5+5)
Phases/tpcc-stock-level/Simple/Explore-10                  254µs ± 1%     250µs ± 1%  -1.74%  (p=0.032 n=4+5)
Phases/tpcc-stock-level/Prepared/Explore-10                102µs ± 3%      98µs ± 0%  -3.29%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-a/Simple/Explore-10        283µs ± 1%     280µs ± 0%  -0.96%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-a/Prepared/Explore-10      257µs ± 0%     255µs ± 0%  -0.77%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-b/Simple/Explore-10        342µs ± 1%     336µs ± 0%  -1.87%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-b/Prepared/Explore-10      292µs ± 1%     290µs ± 0%  -0.70%  (p=0.016 n=5+5)
Phases/many-columns-and-indexes-c/Simple/Explore-10        178ms ± 2%     177ms ± 1%    ~     (p=0.841 n=5+5)
Phases/many-columns-and-indexes-c/Prepared/Explore-10      175ms ± 2%     175ms ± 3%    ~     (p=0.841 n=5+5)
Phases/single-col-histogram-range/Simple/Explore-10       85.0µs ± 0%    85.2µs ± 0%    ~     (p=0.310 n=5+5)
Phases/single-col-histogram-range/Prepared/Explore-10     37.1µs ± 1%    37.7µs ± 2%  +1.62%  (p=0.016 n=5+4)
Phases/batch-insert-one/Simple/Explore-10                 59.0µs ± 0%    58.9µs ± 1%    ~     (p=0.841 n=5+5)
Phases/batch-insert-many/Simple/Explore-10                4.40ms ± 1%    4.35ms ± 0%  -1.02%  (p=0.008 n=5+5)
Phases/ored-preds-100/Simple/Explore-10                   2.52ms ± 1%    2.51ms ± 0%    ~     (p=0.222 n=5+5)
Phases/ored-preds-using-params-100/Simple/Explore-10      2.40ms ± 0%    2.39ms ± 0%    ~     (p=0.056 n=5+5)
Phases/ored-preds-using-params-100/Prepared/Explore-10    1.05ms ± 2%    1.04ms ± 0%  -0.95%  (p=0.008 n=5+5)

name                                                    old alloc/op   new alloc/op   delta
Phases/kv-read/Simple/Explore-10                          15.2kB ± 0%    15.2kB ± 0%    ~     (all equal)
Phases/kv-read-const/Simple/Explore-10                    15.3kB ± 0%    15.3kB ± 0%    ~     (all equal)
Phases/tpcc-new-order/Simple/Explore-10                   32.4kB ± 0%    32.5kB ± 0%  +0.20%  (p=0.008 n=5+5)
Phases/tpcc-delivery/Simple/Explore-10                    26.2kB ± 0%    26.4kB ± 0%  +0.73%  (p=0.008 n=5+5)
Phases/tpcc-delivery/Prepared/Explore-10                  12.6kB ± 0%    12.7kB ± 0%  +1.02%  (p=0.016 n=5+4)
Phases/tpcc-stock-level/Simple/Explore-10                  130kB ± 0%     131kB ± 0%  +0.71%  (p=0.008 n=5+5)
Phases/tpcc-stock-level/Prepared/Explore-10               59.0kB ± 0%    59.6kB ± 0%  +1.11%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-a/Simple/Explore-10       52.0kB ± 0%    52.1kB ± 0%  +0.12%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-a/Prepared/Explore-10     18.0kB ± 0%    18.1kB ± 0%  +0.38%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-b/Simple/Explore-10       60.3kB ± 0%    60.4kB ± 0%  +0.12%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-b/Prepared/Explore-10     20.5kB ± 0%    20.6kB ± 0%  +0.33%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-c/Simple/Explore-10        110MB ± 0%     112MB ± 0%  +1.87%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-c/Prepared/Explore-10      109MB ± 0%     111MB ± 0%  +1.89%  (p=0.008 n=5+5)
Phases/single-col-histogram-range/Simple/Explore-10       29.6kB ± 0%    29.6kB ± 0%    ~     (p=0.397 n=5+5)
Phases/single-col-histogram-range/Prepared/Explore-10     13.0kB ± 0%    13.0kB ± 0%    ~     (all equal)
Phases/batch-insert-one/Simple/Explore-10                 64.9kB ± 0%    64.9kB ± 0%    ~     (p=0.444 n=5+5)
Phases/batch-insert-many/Simple/Explore-10                3.94MB ± 0%    3.94MB ± 0%    ~     (p=0.421 n=5+5)
Phases/ored-preds-100/Simple/Explore-10                    947kB ± 0%     947kB ± 0%    ~     (p=0.421 n=5+5)
Phases/ored-preds-using-params-100/Simple/Explore-10       928kB ± 0%     928kB ± 0%    ~     (p=0.952 n=5+5)
Phases/ored-preds-using-params-100/Prepared/Explore-10     738kB ± 0%     738kB ± 0%    ~     (p=0.841 n=5+5)

name                                                    old allocs/op  new allocs/op  delta
Phases/kv-read/Simple/Explore-10                            92.0 ± 0%      92.0 ± 0%    ~     (all equal)
Phases/kv-read-const/Simple/Explore-10                      92.0 ± 0%      92.0 ± 0%    ~     (all equal)
Phases/tpcc-new-order/Simple/Explore-10                      171 ± 0%       171 ± 0%    ~     (all equal)
Phases/tpcc-delivery/Simple/Explore-10                       161 ± 0%       161 ± 0%    ~     (all equal)
Phases/tpcc-delivery/Prepared/Explore-10                    73.0 ± 0%      73.0 ± 0%    ~     (all equal)
Phases/tpcc-stock-level/Simple/Explore-10                    702 ± 0%       705 ± 0%  +0.43%  (p=0.008 n=5+5)
Phases/tpcc-stock-level/Prepared/Explore-10                  338 ± 0%       341 ± 0%  +0.89%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-a/Simple/Explore-10         95.6 ± 1%      96.0 ± 0%    ~     (p=0.556 n=5+4)
Phases/many-columns-and-indexes-a/Prepared/Explore-10       47.0 ± 0%      47.4 ± 1%    ~     (p=0.444 n=5+5)
Phases/many-columns-and-indexes-b/Simple/Explore-10          151 ± 0%       151 ± 0%    ~     (all equal)
Phases/many-columns-and-indexes-b/Prepared/Explore-10       65.0 ± 0%      65.4 ± 1%    ~     (p=0.556 n=4+5)
Phases/many-columns-and-indexes-c/Simple/Explore-10         750k ± 0%      747k ± 0%  -0.46%  (p=0.008 n=5+5)
Phases/many-columns-and-indexes-c/Prepared/Explore-10       744k ± 0%      741k ± 0%  -0.47%  (p=0.008 n=5+5)
Phases/single-col-histogram-range/Simple/Explore-10          519 ± 0%       519 ± 0%    ~     (all equal)
Phases/single-col-histogram-range/Prepared/Explore-10        250 ± 0%       250 ± 0%    ~     (all equal)
Phases/batch-insert-one/Simple/Explore-10                    326 ± 0%       326 ± 0%    ~     (all equal)
Phases/batch-insert-many/Simple/Explore-10                 28.5k ± 0%     28.5k ± 0%    ~     (p=1.000 n=5+5)
Phases/ored-preds-100/Simple/Explore-10                    4.53k ± 0%     4.53k ± 0%    ~     (all equal)
Phases/ored-preds-using-params-100/Simple/Explore-10       4.53k ± 0%     4.53k ± 0%    ~     (all equal)
Phases/ored-preds-using-params-100/Prepared/Explore-10     2.89k ± 0%     2.89k ± 0%    ~     (all equal)
```

mgartner commented:
*[`pkg/sql/opt/ordering/interesting_orderings.go` line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102011#-NT_0mFx8VOdXtfktfpg:-NT_IQFmBU_qUaMXM-23:b-hyq2iv) ([raw file](https://github.com/cockroachdb/cockroach/blob/779ba3f488d21dbd8081de279d3b7ef404cb4505/pkg/sql/opt/ordering/interesting_orderings.go#L32)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could we do even better by first checking whether any of the orders needs to be simplified? Or does that just not have an impact?
</blockquote></details>

That might be helpful, I didn't try it. It's not really applicable with my latest commit, though, I don't think.


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/props/logical.go` line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102011#-NT_22iDCcAcu4PeixXl:-NT_I-Wi0MEc024RXqUX:b5z72qr) ([raw file](https://github.com/cockroachdb/cockroach/blob/779ba3f488d21dbd8081de279d3b7ef404cb4505/pkg/sql/opt/props/logical.go#L274)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

What's preventing us from simplifying the `InterestingOrderings` field instead of adding a new one?
</blockquote></details>

There's other rules that use un-simplified interesting orderings, like `GeneratePartialOrderTopK` and `GenerateStreamingSetOp`. I didn't check whether or not the behavior of those rules would change if they also use simplified ordering, but I can. 

I think this is less important given my latest commit though.


<!-- Sent from Reviewable.io -->

mgartner commented on pkg/sql/logictest/testdata/logic_test/prepare:
```diff
@@ -1565,7 +1565,6 @@ PREPARE pscs13 AS SELECT $1::bigint, 13
 query T
 SELECT name FROM pg_catalog.pg_prepared_statements ORDER BY name
 ----
-pscs08
```
Unfortunately we may have to update these tests memo sizes change, and it's a bit tedious. I can't think of a good solution for this though...

ankit-gautam23 reviewed (commented):
We need to have a look on test memos size reduction. One way could be clustering.

rharding6373 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice improvements!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102011#-:-NToNr7w3xPFzteHDSjH:bdyovjk)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/logictest/testdata/logic_test/prepare` line 1568 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102011#-NToLn_dBi9hPFal3uUv:-NToLn_dBi9hPFal3uUw:bw5vz15) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec211e70ce48b63173bdc7d7bf7b72c310a923a/pkg/sql/logictest/testdata/logic_test/prepare#L1568)):*
> ```
> SELECT name FROM pg_catalog.pg_prepared_statements ORDER BY name
> ----
> pscs08
> ```

It seems unexpected that some of the successfully prepared statements don't show up in these lists. Is there a reason for this?


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/logictest/testdata/logic_test/prepare` line 1568 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102011#-NToLn_dBi9hPFal3uUv:-NTobIML51YkCT4MaGCU:b-93qrdq) ([raw file](https://github.com/cockroachdb/cockroach/blob/bec211e70ce48b63173bdc7d7bf7b72c310a923a/pkg/sql/logictest/testdata/logic_test/prepare#L1568)):*
This is related to my previous comment:

> Unfortunately we may have to update these tests memo sizes change, and it's a bit tedious. I can't think of a good solution for this though...

These tests are testing the LRU eviction feature @michae2 recently added for the prepared statements cache. Prepared statements in the cache are measured by the estimate size of the memo. Since this PR increase the size of memos, more prepared statements are evicted than before.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 6 of 6 files at r4, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102011#-:-NU3jMZ_5NMT9GN-MM1X:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @michae2 and @rharding6373)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTR!

bors r+
